[Temporal] Implement `with` method for PlainMonthDay

catamorphism opened the pull request:
#### f37c4da1863a1d7e64967e5be68294e1fc0aa29a
<pre>
[Temporal] Implement `with` method for PlainMonthDay

<a href="https://bugs.webkit.org/show_bug.cgi?id=300640">https://bugs.webkit.org/show_bug.cgi?id=300640</a>

Reviewed by Yusuke Suzuki.

with() calls TemporalCalendar::monthDayFromFields(), which calls
TemporalCalendar::isoDateFromFields() -- hence the changes to the type
signature of isoDateFromFields(), which also change the type signatures
of various from() methods.

* JSTests/stress/temporal-plainmonthday.js:
(shouldBe):
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseMonthCode):
(JSC::ISO8601::validMonthCode):
(JSC::ISO8601::monthFromCode): Deleted.
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::calendarResolveFields):
(JSC::TemporalCalendar::isoDateFromFields):
(JSC::TemporalCalendar::monthDayFromFields):
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalCalendarPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::getEpochNanosecondsFor):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::getMonthCode):
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
(JSC::TemporalPlainDate::from):
(JSC::TemporalPlainDate::toPartialDate):
(JSC::TemporalPlainDate::with):
* Source/JavaScriptCore/runtime/TemporalPlainDate.h:
* Source/JavaScriptCore/runtime/TemporalPlainDateConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainDatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.cpp:
(JSC::TemporalPlainDateTime::from):
(JSC::TemporalPlainDateTime::with):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.h:
* Source/JavaScriptCore/runtime/TemporalPlainDateTimeConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDay.cpp:
(JSC::TemporalPlainMonthDay::with):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDay.h:
* Source/JavaScriptCore/runtime/TemporalPlainMonthDayConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSC_DEFINE_CUSTOM_GETTER):

Canonical link: <a href="https://commits.webkit.org/303360@main">https://commits.webkit.org/303360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dce8c5b1764c973037a37a0b2f8ec40e33cac30e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4602 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43130 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139624 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84024 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4549 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4363 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100988 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135055 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118332 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81779 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82843 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124172 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111900 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142270 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130616 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4271 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109361 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4352 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109536 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27753 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3246 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114607 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57515 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4325 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33000 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163583 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4156 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67771 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42516 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4284 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1518,19 +1518,31 @@ String monthCode(uint32_t month)
+    return ParsedMonthCode { monthNumber, isLeapMonth };
+}
+
+bool validMonthCode(StringView monthCode)
```
Why do we need to separate this function and expose it?
Why not just calling `parseMonthCode`?

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.h:
```diff
@@ -375,13 +375,14 @@ String temporalDateToString(PlainDate);
 String temporalMonthDayToString(PlainMonthDay, StringView);
 String monthCode(uint32_t);
-uint8_t monthFromCode(StringView);
+bool validMonthCode(StringView);
```
I think `parseMonthCode` is enough and exposing this function is not necessary.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,49 +234,74 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
 
-        auto otherMonth = ISO8601::monthFromCode(monthCode);
-        if (!otherMonth) {
+        if (!ISO8601::validMonthCode(monthCode.value())) {
```
We should just use `parseMonthCode`.

Constellation replied:
And we should store the result, then.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,49 +234,74 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
-        return { };
+    // Parse monthCode if applicable
+    if (monthCode) {
+        auto otherMonth = ISO8601::parseMonthCode(monthCode.value());
```
We should not parse this again.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,68 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+
 // https://tc39.es/proposal-temporal/#sec-temporal-isodatefromfields
-ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObject, JSObject* temporalDateLike, TemporalOverflow overflow)
+ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObject, JSObject* temporalDateLike, TemporalDateFormat format, std::variant<JSObject*, TemporalOverflow> optionsOrOverflow, TemporalOverflow& overflow)
```
Do not use `std::variant`. We use `Variant`.

catamorphism replied:
Fixed in [2056466](https://github.com/WebKit/WebKit/pull/52251/commits/205646625be51ed6006fc3211b7a35d865ea0406)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,49 +234,74 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
 }
 
-ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObject, double year, double month, double day, TemporalOverflow overflow)
+ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObject, TemporalDateFormat format, double year, double month, double day, std::optional<StringView> monthCode, TemporalOverflow overflow)
```
We should not pass `StringView` monthCode. We should parse it, and pass already parsed ParsedMonthCode.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,68 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
 }
 
+// https://tc39.es/proposal-temporal/#sec-temporal-calendarresolvefields
+static void calendarResolveFields(JSGlobalObject* globalObject, std::optional<double> year, double month, std::optional<StringView> monthCode, TemporalDateFormat format)
```
Ditto.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,68 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        return;
+    }
+    if (monthCode) {
+        auto parsedMonthCode = ISO8601::parseMonthCode(*monthCode);
```
We should not parse it again.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -240,14 +310,33 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
 
+// https://tc39.es/proposal-temporal/#sec-temporal-calendarmonthdayfromfields
+// https://tc39.es/proposal-temporal/#sec-temporal-calendarmonthdaytoisoreferencedate
+ISO8601::PlainDate TemporalCalendar::monthDayFromFields(JSGlobalObject* globalObject, std::optional<double> referenceYear, double month, double day, std::optional<StringView> monthCode, TemporalOverflow overflow)
```
Ditto.

catamorphism replied:
Fixed in [3c2bf05](https://github.com/WebKit/WebKit/pull/52251/commits/3c2bf05645a8abf8e82c2512a4857020021a4ce1)

Constellation commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -766,6 +766,21 @@ void rejectObjectWithCalendarOrTimeZone(JSGlobalObject* globalObject, JSObject*
     }
 }
 
+std::optional<StringView> getMonthCode(JSGlobalObject* globalObject, JSObject* temporalDateLike)
```
Returning `StringView` is wrong as it is not owning lifetime.

catamorphism replied:
It turns out this function wasn't used -- removed it in [724fd55](https://github.com/WebKit/WebKit/pull/52251/commits/724fd5508a49109d531b3c8505c95d2fcc5a75e7).

Constellation reviewed (changes requested):
Overall, I'm very concerning about these `double` year / month / day.
It is very inefficient.
Can you provide what part of the spec makes it double instead of using integers?

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
-    if (!ISO8601::isDateTimeWithinLimits(plainDate.year(), plainDate.month(), plainDate.day(), 12, 0, 0, 0, 0, 0)) {
+    auto plainDate = ISO8601::PlainDate(year, month, day);
+    bool valid = ISO8601::isDateTimeWithinLimits(plainDate.year(), plainDate.month(), plainDate.day(), 12, 0, 0, 0, 0, 0);
+    if (!valid) {
```
Put `[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
     }
 
-    RELEASE_AND_RETURN(scope, isoDateFromFields(globalObject, year, month, day, overflow));
+    if (!(day >= 1 && day <= daysInMonth)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+        day = std::min<unsigned>(day, daysInMonth);
+    }
+
+    if (!(month >= 1 && month <= 12)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
-        throwTypeError(globalObject, scope, "year property must be present"_s);
+    // Duplicate code from TemporalPlainDate::toPlainDate so we can convert from
+    // double to int32_t / unsigned here
+    if (!ISO8601::isYearWithinLimits(year)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+        ASSERT(otherMonth);
+
+        // FIXME: ISO8601 calendar assumed
+        if (otherMonth->monthNumber < 1 || otherMonth->monthNumber > 12 || otherMonth->isLeapMonth) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+    if (!yearProperty.isUndefined()) {
+        year = yearProperty.toIntegerOrInfinity(globalObject);
+        RETURN_IF_EXCEPTION(scope, { });
+        if (!std::isfinite(year)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (format != TemporalDateFormat::MonthDay) {
+        if (yearProperty.isUndefined()) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
-        auto otherMonth = ISO8601::monthFromCode(monthCode);
+        otherMonth = ISO8601::parseMonthCode(monthCodeString);
+
         if (!otherMonth) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -178,6 +219,8 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+    std::optional<ParsedMonthCode> otherMonth;
+    bool monthCodePresent = false;
     if (monthCodeProperty.isUndefined()) {
         if (monthProperty.isUndefined()) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,65 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        if (!dayProperty.isUndefined()) {
+            day = dayProperty.toIntegerOrInfinity(globalObject);
+            RETURN_IF_EXCEPTION(scope, { });
+            if (!(day > 0 && std::isfinite(day))) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,65 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        JSValue dayProperty = temporalDateLike->get(globalObject, vm.propertyNames->day);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (dayProperty.isUndefined()) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,65 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        if (month != monthCode->monthNumber) {
+            throwRangeError(globalObject, scope, "month does not match month code"_s);
+            return;
+        }
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -148,24 +148,65 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    // FIXME: non-ISO8601 calendars
+    if ((format == TemporalDateFormat::Date || format == TemporalDateFormat::YearMonth)
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+    }
+
+    if (monthCodePresent) {
+        if (!otherMonth) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+        }
+        if (!month)
+            month = otherMonth->monthNumber;
+        else if (month.value() != otherMonth->monthNumber) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+
+    // Duplicate code from TemporalPlainDate::toPlainDate so we can convert from
+    // double to int32_t / unsigned here
+    if (year && !ISO8601::isYearWithinLimits(*year)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+        if (overflow == TemporalOverflow::Constrain)
+            monthToUse = std::min<unsigned>(*month, 12);
+        else {
+            if (!(month >= 1 && month <= 12)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+
+    if (overflow == TemporalOverflow::Constrain)
+        dayToUse = std::min<unsigned>(dayToUse, daysInMonth);
+    else if (!(dayToUse >= 1 && dayToUse <= daysInMonth)) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -284,23 +339,15 @@ ISO8601::PlainDate TemporalPlainDate::with(JSGlobalObject* globalObject, JSObjec
+    auto [y, m, d, optionalMonthCode, overflow, any] = mergeDateFields(globalObject, temporalDateLike, optionsValue, year(), month(), day());
     RETURN_IF_EXCEPTION(scope, { });
-    if (!optionalYear && !optionalMonth && !optionalDay) {
+    if (any == TemporalAnyProperties::None) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDateTimeConstructor.cpp:
```diff
@@ -119,15 +119,16 @@ JSC_DEFINE_HOST_FUNCTION(temporalPlainDateTimeConstructorFuncFrom, (JSGlobalObje
-    if (itemValue.inherits<TemporalPlainDateTime>())
+    if (itemValue.inherits<TemporalPlainDateTime>()) {
+        // Validate overflow
+        toTemporalOverflow(globalObject, options);
```
Error check is missing after this call.

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainMonthDay.cpp:
```diff
@@ -118,6 +119,32 @@ String TemporalPlainMonthDay::toString(JSGlobalObject* globalObject, JSValue opt
+    rejectObjectWithCalendarOrTimeZone(globalObject, temporalMonthDayLike);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (!calendar()->isISO8601()) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainMonthDay.cpp:
```diff
@@ -118,6 +119,32 @@ String TemporalPlainMonthDay::toString(JSGlobalObject* globalObject, JSValue opt
+        TemporalPlainDate::mergeDateFields(globalObject, temporalMonthDayLike, optionsValue,
+            1972, month(), day());
+    RETURN_IF_EXCEPTION(scope, { });
+    if (any == TemporalAnyProperties::None) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainMonthDayPrototype.cpp:
```diff
@@ -126,6 +128,28 @@ JSC_DEFINE_HOST_FUNCTION(temporalPlainMonthDayPrototypeFuncToLocaleString, (JSGl
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto* monthDay = jsDynamicCast<TemporalPlainMonthDay*>(callFrame->thisValue());
+    if (!monthDay)
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainMonthDayPrototype.cpp:
```diff
@@ -126,6 +128,28 @@ JSC_DEFINE_HOST_FUNCTION(temporalPlainMonthDayPrototypeFuncToLocaleString, (JSGl
+        return throwVMTypeError(globalObject, scope, "Temporal.PlainMonthDay.prototype.with called on value that's not a PlainMonthDay"_s);
+
+    JSValue temporalMonthDayLike  = callFrame->argument(0);
+    if (!temporalMonthDayLike.isObject())
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
```diff
@@ -266,9 +270,60 @@ std::array<std::optional<double>, numberOfTemporalPlainDateUnits> TemporalPlainD
+    }
+
+    TemporalOverflow overflow = toTemporalOverflow(globalObject, optionsValue);
+
```
Remove this empty line.

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -189,65 +232,129 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
+    RETURN_IF_EXCEPTION(scope, { });
+    double year = referenceYear.value_or(1972);
+    auto result = TemporalDuration::regulateISODate(year, month, day, overflow);
+    if (!result || !ISO8601::isValidISODate(result->year(), result->month(), result->day())) {
```
`[[unlikely]]`

catamorphism replied:
Fixed in [348c785](https://github.com/WebKit/WebKit/pull/52251/commits/348c785580dc01981e23b15ddbfdafabda503920)

Constellation reviewed (commented):
Commented. Please check all error cases are annotated as `[[unlikely]]` and checking whether there is missing error cases.

catamorphism committed:
[Temporal] Implement `with` method for PlainMonthDay

https://bugs.webkit.org/show_bug.cgi?id=300640

Reviewed by Yusuke Suzuki.

with() calls TemporalCalendar::monthDayFromFields(), which calls
TemporalCalendar::isoDateFromFields() -- hence the changes to the type
signature of isoDateFromFields(), which also change the type signatures
of various from() methods.

* JSTests/stress/temporal-plainmonthday.js:
(shouldBe):
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseMonthCode):
(JSC::ISO8601::validMonthCode):
(JSC::ISO8601::monthFromCode): Deleted.
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::calendarResolveFields):
(JSC::TemporalCalendar::isoDateFromFields):
(JSC::TemporalCalendar::monthDayFromFields):
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalCalendarPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::getEpochNanosecondsFor):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::getMonthCode):
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
(JSC::TemporalPlainDate::from):
(JSC::TemporalPlainDate::toPartialDate):
(JSC::TemporalPlainDate::with):
* Source/JavaScriptCore/runtime/TemporalPlainDate.h:
* Source/JavaScriptCore/runtime/TemporalPlainDateConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainDatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.cpp:
(JSC::TemporalPlainDateTime::from):
(JSC::TemporalPlainDateTime::with):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.h:
* Source/JavaScriptCore/runtime/TemporalPlainDateTimeConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDay.cpp:
(JSC::TemporalPlainMonthDay::with):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDay.h:
* Source/JavaScriptCore/runtime/TemporalPlainMonthDayConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainMonthDayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSC_DEFINE_CUSTOM_GETTER):

Canonical link: https://commits.webkit.org/303360@main
